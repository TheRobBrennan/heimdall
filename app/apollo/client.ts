import fetch from "cross-fetch"
import { useMemo } from "react"

import { ApolloClient, createHttpLink, InMemoryCache } from "@apollo/client"
import { setContext } from "@apollo/client/link/context"

let apolloClient

const httpLink = createHttpLink({
  uri: `api/graphql`,
  fetch,
})

/* istanbul ignore next */
const authLink = setContext(async () => {
  // See https://www.apollographql.com/docs/react/networking/authentication/#header for more details
  // Get our session detail from the back-end API
  let authorization = ""
  try {
    const data = await (await fetch("api/session")).json()
    const { accessToken } = data
    authorization = `Bearer ${accessToken}`
  } catch (e) {
    // No session data available; no authorization header will be sent
  }

  // Return the headers to the context so httpLink can read them
  return {
    headers: {
      authorization,
    },
  }
})

function createApolloClient(initialState) {
  const ssrMode = typeof window === "undefined"

  return new ApolloClient({
    ssrMode,
    link: authLink.concat(httpLink),
    cache: new InMemoryCache().restore(initialState),
  })
}

export function initializeApollo(initialState = null) {
  const _apolloClient = apolloClient ?? createApolloClient(initialState)

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // get hydrated here
  if (initialState) {
    _apolloClient.cache.restore(initialState)
  }
  // For SSG and SSR always create a new Apollo Client
  /* istanbul ignore next */
  if (typeof window === "undefined") return _apolloClient
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient

  return _apolloClient
}

/* istanbul ignore next */
export function useApollo(initialState) {
  const store = useMemo(() => initializeApollo(initialState), [initialState])
  return store
}
