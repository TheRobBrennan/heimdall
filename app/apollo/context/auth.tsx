// Source https://github.com/johnymontana/grandcast.fm/blob/master/next-app/lib/auth.js
import React, { useState, useContext, createContext } from "react"
import {
  ApolloProvider,
  ApolloClient,
  InMemoryCache,
  HttpLink,
  // gql, /* REVISIT: Import gql when auth mutations are ready to be invoked */
} from "@apollo/client"
import fetch from "cross-fetch"

const AuthContext = createContext(null)

/* istanbul ignore next */
export const useAuth = () => {
  return useContext(AuthContext)
  // Returns the object from an instantiated useProvideAuth function
  // (e.g. {
  //   createApolloClient,
  //   initializeApollo,
  //   signIn,
  //   signOut,
  //   isSignedIn,
  //  }
  // )
}

export function AuthProvider({ children = null }) {
  const auth = useProvideAuth()

  return (
    <AuthContext.Provider value={auth}>
      <ApolloProvider client={auth.createApolloClient(null)}>
        {children}
      </ApolloProvider>
    </AuthContext.Provider>
  )
}

/* istanbul ignore next */
function useProvideAuth() {
  let apolloClient

  const [authToken, setAuthToken] = useState(null)

  const getAuthHeaders = () => {
    if (!authToken) return null

    return {
      authorization: `Bearer ${authToken}`,
    }
  }

  function createApolloClient(initialState) {
    const ssrMode = typeof window === "undefined"

    const link = new HttpLink({
      uri: "api/graphql",
      headers: getAuthHeaders(),
      fetch,
    })

    return new ApolloClient({
      ssrMode,
      link,
      cache: new InMemoryCache().restore(initialState),
    })
  }

  function initializeApollo(initialState = null) {
    const _apolloClient = apolloClient ?? createApolloClient(initialState)

    // If your page has Next.js data fetching methods that use Apollo Client, the initial state
    // get hydrated here
    if (initialState) {
      _apolloClient.cache.restore(initialState)
    }
    // For SSG and SSR always create a new Apollo Client
    /* istanbul ignore next */
    if (typeof window === "undefined") return _apolloClient
    // Create the Apollo Client once in the client
    if (!apolloClient) apolloClient = _apolloClient

    return _apolloClient
  }

  const signIn = async ({ username, password }) => {
    // REVISIT: Implement signIn login mutation and set authToken
    console.log(
      `(!) TODO: Implement signIn login mutation and set authToken for username: ${username} :: password: ${password} (!)`
    )
    // const client = createApolloClient(null)
    // const LoginMutation = gql`
    //   mutation LoginMutation($username: String!, $password: String!) {
    //     login(username: $username, password: $password) {
    //       token
    //     }
    //   }
    // `
    // const result = await client.mutate({
    //   mutation: LoginMutation,
    //   variables: { username, password },
    // })
    // console.log(result)
    // if (result?.data?.login?.token) {
    //   setAuthToken(result.data.login.token)
    // }
  }

  const signOut = () => {
    // REVISIT: Implement signOut once authToken has been set
    console.log(`(!) TODO: Implement signOut once authToken has been set (!)`)
    // setAuthToken(null)
  }

  const isSignedIn = () => {
    // REVISIT: Implement isSignedIn once authToken has been set
    console.log(
      `(!) TODO: Implement isSignedIn once authToken has been set (!)`
    )
    // if (authToken) {
    //   return true
    // } else {
    //   return false
    // }
  }

  return {
    createApolloClient,
    initializeApollo,
    signIn,
    signOut,
    isSignedIn,
  }
}
