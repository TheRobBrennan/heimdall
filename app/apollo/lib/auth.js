// Source https://github.com/johnymontana/grandcast.fm/blob/master/next-app/lib/auth.js
import React, { useState, useContext, createContext } from "react"
import {
  ApolloProvider,
  ApolloClient,
  InMemoryCache,
  HttpLink,
  gql,
} from "@apollo/client"
import fetch from "cross-fetch"

const authContext = createContext()

export const useAuth = () => {
  return useContext(authContext)
}

function useProvideAuth() {
  let apolloClient

  const [authToken, setAuthToken] = useState(null)

  const getAuthHeaders = () => {
    if (!authToken) return null

    return {
      authorization: `Bearer ${authToken}`,
    }
  }

  function createApolloClient(initialState) {
    const ssrMode = typeof window === "undefined"

    const link = new HttpLink({
      uri: "api/graphql",
      headers: getAuthHeaders(),
      fetch,
    })

    return new ApolloClient({
      ssrMode,
      link,
      cache: new InMemoryCache().restore(initialState),
    })
  }

  function initializeApollo(initialState = null) {
    const _apolloClient = apolloClient ?? createApolloClient(initialState)

    // If your page has Next.js data fetching methods that use Apollo Client, the initial state
    // get hydrated here
    if (initialState) {
      _apolloClient.cache.restore(initialState)
    }
    // For SSG and SSR always create a new Apollo Client
    /* istanbul ignore next */
    if (typeof window === "undefined") return _apolloClient
    // Create the Apollo Client once in the client
    if (!apolloClient) apolloClient = _apolloClient

    return _apolloClient
  }

  const signOut = () => {
    setAuthToken(null)
  }

  const signIn = async ({ username, password }) => {
    const client = createApolloClient()
    const LoginMutation = gql`
      mutation LoginMutation($username: String!, $password: String!) {
        login(username: $username, password: $password) {
          token
        }
      }
    `
    const result = await client.mutate({
      mutation: LoginMutation,
      variables: { username, password },
    })

    console.log(result)

    if (result?.data?.login?.token) {
      setAuthToken(result.data.login.token)
    }
  }

  const isSignedIn = () => {
    if (authToken) {
      return true
    } else {
      return false
    }
  }

  return {
    createApolloClient,
    initializeApollo,
    signIn,
    signOut,
    isSignedIn,
  }
}

export function AuthProvider({ children }) {
  const auth = useProvideAuth()

  return (
    <authContext.Provider value={auth}>
      <ApolloProvider client={auth.createApolloClient()}>
        {children}
      </ApolloProvider>
    </authContext.Provider>
  )
}
