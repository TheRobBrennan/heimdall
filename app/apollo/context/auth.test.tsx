import * as React from "react"
import { cleanup, render } from "@testing-library/react"
import { MockedProvider } from "@apollo/client/testing"

import { AuthProvider as Component } from "./auth"

describe("Our AuthProvider component", () => {
  afterEach(cleanup)

  it("should render without children", () => {
    const { asFragment } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <Component />
      </MockedProvider>
    )

    expect(asFragment).toMatchSnapshot()
  })
  it("should render with children", () => {
    const { asFragment } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <Component>
          <h1>An example heading</h1>
        </Component>
      </MockedProvider>
    )

    expect(asFragment).toMatchSnapshot()
  })
})
