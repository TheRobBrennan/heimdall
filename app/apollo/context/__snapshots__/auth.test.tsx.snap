// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Our AuthProvider component should render with children 1`] = `[Function]`;

exports[`Our AuthProvider component should render without children 1`] = `[Function]`;
